File: sample-conference-videochat/sample-conference-videochat/Components/Calls/CallViewController/OpponentsFlowLayout.m
//
//  CustomLayout.m
//  flowlayout
//
//  Created by Anton Sokolchenko on 10/8/15.
//  Copyright © 2015 anton. All rights reserved.
//

#import "OpponentsFlowLayout.h"

@interface OpponentsFlowLayout()
{
    NSMutableArray *_layoutAttributes;
}

@end

@implementation OpponentsFlowLayout

// MARK: Construction

- (instancetype)init {
    
    self = [super init];
    if (self != nil) {
        
        [self commonInit];
    }
    
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    
    self = [super initWithCoder:aDecoder];
    if (self != nil) {
        
        [self commonInit];
    }
    
    return self;
}

- (void)commonInit {
    
    self.minimumInteritemSpacing = 2;
    self.minimumLineSpacing = 2;
    _layoutAttributes = [[NSMutableArray alloc] init];
}

// MARK: UISubclassingHooks

- (void)prepareLayout {
    [_layoutAttributes removeAllObjects];
    
    NSInteger numberOfItems = [self.collectionView numberOfItemsInSection:0];
    for (NSInteger i = 0; i < numberOfItems; ++i) {
        UICollectionViewLayoutAttributes *attributes = [UICollectionViewLayoutAttributes layoutAttributesForCellWithIndexPath:[NSIndexPath indexPathForItem:i inSection:0]];
        attributes.frame = [self itemFrameWithItemIndex:i itemsCount:numberOfItems];
        [_layoutAttributes addObject:attributes];
    }
}

- (CGSize)collectionViewContentSize {
    
    return self.collectionView.frame.size;
}

- (NSArray *)layoutAttributesForElementsInRect:(CGRect)rect {
    NSMutableArray *array = [[NSMutableArray alloc] init];
    for (UICollectionViewLayoutAttributes *attributes in _layoutAttributes) {
        if (!CGRectIsNull(CGRectIntersection(rect, attributes.frame))) {
            [array addObject:attributes];
        }
    }
    return array;
}

- (BOOL)shouldInvalidateLayoutForBoundsChange:(CGRect)newBounds {
    return YES;
}

// MARK: Helpers

NSInteger shiftPosition(NSUInteger itemsCount, BOOL isPortrait) {
    // ItemsCount : position
    NSDictionary *map = nil;
    if (isPortrait) {
        map = @
        {
            @3 : @0,
            @5 : @0,
            @7 : @6,
            @8 : @6,
            @10 : @9,
            @11 : @9,
            @13 : @12,
            @14 : @12,
        };
    }
    else {
        map = @
        {
            @5 : @3,
            @7 : @4,
            @10 : @8,
        };
    }
    
    NSNumber *position = map[@(itemsCount)];
    
    if (position) {
        return position.integerValue;
    }
    
    return -1;
}

- (CGRect)itemFrameWithItemIndex:(NSUInteger)itemIndex
                      itemsCount:(NSUInteger)itemsCount {
    
    CGSize contentSize = [self collectionViewContentSize];
    BOOL isPortrait = contentSize.width < contentSize.height;
    NSInteger columns = numberOfColumns(itemsCount, isPortrait);
    
    if (itemsCount > 1) {
        
        NSInteger shiftPos = shiftPosition(itemsCount, isPortrait);
        NSInteger shift = itemIndex > shiftPos;
        
        CGFloat rows = ceilf((CGFloat)itemsCount / columns);
        NSUInteger mod = itemsCount % columns;
        
        CGFloat scale = 1.0f/columns;
        if (shiftPos == itemIndex) {
            if (columns == 2) {
                scale = 1.0f;
            }
            else if (columns == 3) {
                scale = mod == 1 ? 1.0f : (CGFloat)mod/columns;
            }
            else if (columns == 4) {
                scale = 2.0f/columns;
            }
        }
        
        CGFloat w = contentSize.width * scale;
        CGFloat h = contentSize.height / rows;
        CGFloat i = (itemIndex + shift);
        
        CGFloat row = floorf(i / columns);
        NSUInteger col = ((itemIndex + shift) % columns);
        
        return CGRectMake(w * col, h * row, w, h);
    }
    else {
        
        return CGRectMake(0, 0, contentSize.width, contentSize.height);
    }
}

static NSInteger numberOfColumns(NSInteger numberOfItems, BOOL isPortrait) {
    
    NSInteger countOfColumns;
    if (isPortrait) {
        switch (numberOfItems) {
            case 1:
            case 2:
                countOfColumns = 1;
                break;
            case 3:
            case 4:
            case 5:
            case 6:
                countOfColumns = 2;
                break;
            default:
                countOfColumns = 3;
                break;
        }
    }
    else {
        switch (numberOfItems) {
            case 1:
                countOfColumns = 1;
                break;
            case 2:
            case 4:
                countOfColumns = 2;
                break;
            case 3:
            case 5:
            case 6:
            case 9:
                countOfColumns = 3;
                break;
            default:
                countOfColumns = 4;
                break;
        }
    }
    return countOfColumns;
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Calls/CallViewController/OpponentsFlowLayout.swift
//
//  OpponentsFlowLayout.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 04.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit

class OpponentsFlowLayout: UICollectionViewFlowLayout {
    
    private var layoutAttributes = [UICollectionViewLayoutAttributes]()
    
    // MARK: Construction
    override init() {
        super.init()
        commonInit()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        commonInit()
    }
    
    func commonInit() {
        
        minimumInteritemSpacing = 2
        minimumLineSpacing = 2
        
    }
    // MARK: UISubclassingHooks
    
    override func prepare() {
        layoutAttributes.removeAll()
        
        let numberOfItems: Int? = collectionView?.numberOfItems(inSection: 0)
        for i in 0..<(numberOfItems ?? 0) {
            let attributes = UICollectionViewLayoutAttributes(forCellWith: IndexPath(item: i, section: 0))
            attributes.frame = itemFrame(withItemIndex: i, itemsCount: numberOfItems!)
            layoutAttributes.append(attributes)
        }
    }
    
    override var collectionViewContentSize: CGSize {
        
        return collectionView?.frame.size ?? CGSize.zero
    }
    
    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        var array: [UICollectionViewLayoutAttributes] = []
        for attributes in layoutAttributes {
            if !rect.intersection((attributes.frame)).isNull {
    
                    array.append(attributes)

            }
        }
        return array
    }
    
    override func shouldInvalidateLayout(forBoundsChange newBounds: CGRect) -> Bool {
        return true
    }
    
    func shiftPosition(itemsCount: Int, isPortrait: Bool) -> Int {
        // ItemsCount : position
        var map: [Int: Int]? = nil
        if isPortrait {
            map = [3: 0, 5: 0, 7: 6, 8: 6, 10: 9, 11: 9, 13: 12, 14: 12]
        } else {
            map = [5: 3, 7: 4, 10: 8]
        }
        
        if let position = map?[itemsCount] {
            return position
        }

        return -1
    }
    
    func itemFrame(withItemIndex itemIndex: Int, itemsCount: Int) -> CGRect {
        
        let contentSize: CGSize = collectionViewContentSize
        let isPortrait: Bool = contentSize.width < contentSize.height
        let columns = numberOfColumns(numberOfItems: itemsCount, isPortrait: isPortrait)
        
        if itemsCount > 1 {
            
            let shiftPos = shiftPosition(itemsCount: itemsCount, isPortrait: isPortrait)
            let shift = itemIndex > shiftPos ? 1 : 0
            
            let rows: CGFloat = CGFloat(ceilf(Float(itemsCount/columns)))
            let mod: Int = itemsCount % columns
            
            var scale: CGFloat = 1.0 / CGFloat(columns)
            if shiftPos == itemIndex {
                if columns == 2 {
                    scale = 1.0
                } else if columns == 3 {
                    scale = mod == 1 ? 1.0 : CGFloat(mod) / CGFloat(columns)
                } else if columns == 4 {
                    scale = 2.0 / CGFloat(columns)
                }
            }
            
            let w: CGFloat = contentSize.width * scale
            let h: CGFloat = contentSize.height / rows
            let i = CGFloat((itemIndex + shift))
            
            let row = CGFloat(floorf(Float(i / CGFloat(columns))))
            let col: Int = (itemIndex + shift) % columns
            
            return CGRect(x: w * CGFloat(col), y: h * row, width: w, height: h)
        } else {
            
            return CGRect(x: 0, y: 0, width: contentSize.width, height: contentSize.height)
        }
    }
    
    private func numberOfColumns(numberOfItems: Int, isPortrait: Bool) -> Int {
        
        var countOfColumns: Int
        if isPortrait {
            switch numberOfItems {
            case 1, 2:
                countOfColumns = 1
            case 3, 4, 5, 6:
                countOfColumns = 2
            default:
                countOfColumns = 3
            }
        } else {
            switch numberOfItems {
            case 1:
                countOfColumns = 1
            case 2, 4:
                countOfColumns = 2
            case 3, 5, 6, 9:
                countOfColumns = 3
            default:
                countOfColumns = 4
            }
        }
        return countOfColumns
    }
}
